README
Description
-----------
This is a CardDAV-Implementation for roundcube 0.6 or higher.

Fork of christian-putzke/Roundcube-CardDAV 
Changes:
--------
-vcard cleanup (replace ITEMX.ADR by ADR)
-Update french translation
-Ability to use Roundcube session password for CardDav authentication
   if carddav server and imap server share the same password, password changes no need to modify carddav preferences
   in password field, "%p" is dynamicaly replaced by roundcube session password

Features
--------
* Add multiple CardDAV server for each user
* CardDAV contacts are stored in the local database which provides great performance
* Tested CardDAV servers: DAViCal, Apple Addressbook Server, meetoo, SabreDAV, ownCloud, SOGo
* You can read / add / delete / edit CardDAV contacts (vCards)
* Autocomplete all CardDAV contacts within the compose email view
* Search for all CardDAV contacts within the addressbook
* Automaticly synchronized CardDAV contacts (just execute /plugins/carddav/cronjob/synchronize.php within the crontab)


Planned features
----------------
* Improved search for CardDAV contacts within the addressbook


Requirements
------------
* MySQL or PostgreSQL
* CURL


Installation
------------
* Execute SQL statements from /plugins/carddav/SQL/yourDatabase.sql
* Add 'carddav' to the plugins array in /config/main.inc.php
* Copy /plugins/carddav/config.inc.php.dist to /plugins/carddav/config.inc.php
* Login into your roundcube webmail and add your CardDAV server in the settings


Update
------
* Execute new SQL statements from /plugins/carddav/SQL/yourDatabase.update.sql


Special thanks
--------------
* B5r1oJ0A9G for the PostgreSQL statements!


CardDAV server list
-------------------
* DAViCal: https://example.com/{resource|principal|username}/{collection}/
* Apple Addressbook Server: https://example.com/addressbooks/users/{resource|principal|username}/{collection}/
* memotoo: https://sync.memotoo.com/cardDAV/
* SabreDAV: https://example.com/addressbooks/{resource|principal|username}/{collection}/
* ownCloud: https://example.com/remote.php/carddav/addressbooks/{resource|principal|username}/{collection}/
* SOGo: https://example.com/SOGo/dav/{resource|principal|username}/Contacts/{collection}/


Contact
-------
* Author: Christian Putzke <christian.putzke@graviox.de>
* Report feature requests and bugs here: https://github.com/graviox/Roundcube-CardDAV/issues
* Visit Graviox Studios: http://www.graviox.de/
* Follow me on Twitter: https://twitter.com/graviox/


CHANGELOG
Changes from v0.5.2 to v0.5.3
- At least for the Apple's Calendar Server the custom HTTP request PROPFIND in combination with
  infinity depth is no longer allowed. Now, the depth is always specified as 1.
- Fixed the calls to parse_input_value by the static function rcube_utils::parse_input_value.
- Made write_log function of carddav static.

Changes from v0.5.1 to v0.5.2
- Replaced the deprecated rcmail function get_table_name.

Changes from v0.5 to v0.5.1
- Added PostgreSQL support (Thanks to B5r1oJ0A9G for the PostgreSQL statements!)

Changes from v0.4 to v0.5
- Added automaticly synchronized CardDAV contacts via cronjob
- Added larry skin support
- Added list of CardDAV server URLs
- Added read only option for CardDAV servers
- Added SOGo support
- Added package.xml
- Major standard skin UI improvements
- CardDAV backend class update to v0.5.1
- Minor comment, phpdoc and documentation changes

Changes from v0.3.1 to v0.4
- Added add functionality for CardDAV contacts
- Added vCard import functionality for CardDAV addressbooks
- License change from LGPLv2 to AGPLv3
- CardDAV backend class update to v0.4.9
- Added logging functionality
- Each CardDAV server is now an own addressbook not a group of a global CardDAV addressbook like before
- Minor bugfixes
- Improved synchronization

Changes from v0.3 to v0.3.1
- Bugfix: didn't merged vCards on edit correctly

Changes from v0.2.5 to v0.3
- Added edit and delete functionality for CardDAV contacts
- Added sabreDAV support
- Added ownCloud support
- CardDAV backend class update to v0.4.6
- Restructured and cleaned carddav_addressbook class (not completly done yet)
- Minor improvements (CURL install check, usability improvements, sync don't add empty vCards now, ...)
- Added IT and FR language files

Changes from v0.2.4 to v0.2.5
- Bugfix: email and name can now be null
- Ajax POST contents are now base64 encoded
- Username and password can now be empty
- Minor localization changes

Changes from v0.2.3 to v0.2.4
- Bugfix: last_modified date is now STRING instead of INT
- CardDAV backend class update to v0.4.2

Changes from v0.2.2 to v0.2.3
- Added memotoo support
- CardDAV-Backend class update to v0.4.1
- Added addressbook search functionality
- Changed version naming
- Restructured carddav_addressbook class

Changes from 0.2.1 to 0.2.2
- CardDAV backend class update to v0.4
- Added apple addressbook server support

Changes from 0.2 to 0.2.1
- Show CardDAV contacts list (addressbook pagination) bugfix
- Autocomplete CardDAV contacts groups bugfix
- CardDAV groups will now be displayed correctly in the CardDAV contact detail view
- Hide CardDAV contacts sync button if no CardDAV servers were added
- Fallback skin path added "skins/default"
- Minor localization changes

Changes from 0.1 to 0.2
- Save / delete CardDAV server settings via ajax
- Minor class structure changes
- Minor localization changes
- Minor database changes
- Initial CardDAV contacts sync
- Addressbook integration (read only)
- Manuel CardDAV contacts sync (read only)
- CardDAV backend class update to v0.3.4
- Autocomplete CardDAV contacts

release 0.1
- Save / delete CardDAV server settings
- Realtime CardDAV server check
- Multiple CardDAV server for each user
- English and German localization


carddav_settings.js
/**
 * Roundcube CardDAV addressbook extension
 *
 * @author Christian Putzke <christian.putzke@graviox.de>
 * @copyright Christian Putzke @ Graviox Studios
 * @since 12.09.2011
 * @link http://www.graviox.de/
 * @link https://twitter.com/graviox/
 * @version 0.5.1
 * @license http://www.gnu.org/licenses/agpl.html GNU AGPL v3 or later
 *
 */

if (window.rcmail)
{
	rcmail.addEventListener('init', function(evt)
	{
		var tab = $('<span>').attr('id', 'settingstabplugincarddav-server').addClass('tablink');

		var button = $('<a>').attr('href', rcmail.env.comm_path+'&_action=plugin.carddav-server').
			html(rcmail.gettext('settings_tab', 'carddav')).appendTo(tab);
		
		rcmail.add_element(tab, 'tabs');
		rcmail.addEventListener('plugin.carddav_server_message', carddav_server_message);

		rcmail.register_command('plugin.carddav-server-save', carddav_server_add, true);

		rcmail.register_command('plugin.carddav-server-delete', function(carddav_server_id)
		{
			rcmail.http_post(
				'plugin.carddav-server-delete',
				'_carddav_server_id=' + $.base64Encode(carddav_server_id),
				rcmail.display_message(rcmail.gettext('settings_delete_loading', 'carddav'), 'loading')
			);
		}, true);

		$('#_label').keypress(carddav_server_add_enter_event);
		$('#_server_url').keypress(carddav_server_add_enter_event);
		$('#_username').keypress(carddav_server_add_enter_event);
		$('#_password').keypress(carddav_server_add_enter_event);
	});

	function carddav_server_add_enter_event(e)
	{
		if (e.keyCode == 13)
		{
			carddav_server_add();
		}
	};

	function carddav_server_add()
	{
		var input_label		= rcube_find_object('_label');
		var input_url		= rcube_find_object('_server_url');
		var input_username	= rcube_find_object('_username');
		var input_password	= rcube_find_object('_password');
		var input_read_only	= rcube_find_object('_read_only');

		if (input_label.value == '' || input_url.value == '')
		{
			rcmail.display_message(rcmail.gettext('settings_empty_values', 'carddav'), 'error');
		}
		else
		{
			rcmail.http_post(
				'plugin.carddav-server-save',
				'_label=' + $.base64Encode(input_label.value) + '&_server_url=' + $.base64Encode(input_url.value) + '&_username=' + $.base64Encode(input_username.value) + '&_password=' + $.base64Encode(input_password.value)  + '&_read_only=' + $.base64Encode(input_read_only.checked === true ? '1' : '0'),
				rcmail.display_message(rcmail.gettext('settings_init_server', 'carddav'), 'loading')
			);
		}
	}

	function carddav_server_message(response)
	{
		if (response.check)
		{
			$('#carddav_server_list').slideUp();
			$('#carddav_server_list').html(response.server_list);
			$('#carddav_server_list').slideDown();

			rcmail.display_message(response.message, 'confirmation');
		}
		else
		{
			rcmail.display_message(response.message, 'error');
		}
	}
}

carddav_addressbook.js
/**
 * Roundcube CardDAV addressbook extension
 *
 * @author Christian Putzke <christian.putzke@graviox.de>
 * @copyright Christian Putzke @ Graviox Studios
 * @since 22.09.2011
 * @link http://www.graviox.de/
 * @link https://twitter.com/graviox/
 * @version 0.5.1
 * @license http://www.gnu.org/licenses/agpl.html GNU AGPL v3 or later
 *
 */

if (window.rcmail)
{
	rcmail.addEventListener('init', function(evt)
	{
		rcmail.enable_command('plugin.carddav-addressbook-sync', true);
		rcmail.addEventListener('plugin.carddav_addressbook_message', carddav_addressbook_message);
		
		rcmail.register_command('plugin.carddav-addressbook-sync', function()
		{
			rcmail.http_post(
				'plugin.carddav-addressbook-sync',
				'',
				rcmail.display_message(rcmail.gettext('addressbook_sync_loading', 'carddav'), 'loading')
			);
		}, true);
	});
	
	function carddav_addressbook_message(response)
	{
		if (response.check)
		{
			rcmail.display_message(response.message, 'confirmation');
		}
		else
		{
			rcmail.display_message(response.message, 'error');
		}
	}
	
}

cronjob/synchronize.php
<?php

/**
 * Roundcube CardDAV synchronization
 *
 * @author Christian Putzke <christian.putzke@graviox.de>
 * @copyright Christian Putzke @ Graviox Studios
 * @since 31.03.2012
 * @link http://www.graviox.de/
 * @link https://twitter.com/graviox/
 * @version 0.5
 * @license http://www.gnu.org/licenses/agpl.html GNU AGPL v3 or later
 *
 */
class carddav_synchronization_cronjob
{
	/**
	 * @var	string	$doc_root	Roundcubes document root
	 */
	private	$doc_root;

	/**
	 * @var	string	$rc_inset	Roundcubes iniset script
	 */
	private	$rc_iniset = 'program/include/iniset.php';

	/**
	 * Init CardDAV synchronization cronjob
	 *
	 * @param	boolean	$init	Initialization
	 */
	public function __construct($init = true)
	{
		if ($init === true)
		{
			$this->init();
		}
	}

	/**
	 * Init CardDAV synchronization cronjob
	 *
	 * @return	void
	 */
	public function init()
	{
		$this->detect_document_root();
		$this->include_rc_iniset();
	}

	/**
	 * Detect Roundcubes real document root
	 *
	 * @return	void
	 */
	private function detect_document_root()
	{
		$dir = dirname(__FILE__);
		$dir = str_replace('plugins\\carddav\\cronjob', null, $dir);
		$this->doc_root = str_replace('plugins/carddav/cronjob', null, $dir);
		define('INSTALL_PATH', $this->doc_root);
	}

	/**
	 * Include Roundcubes initset so that the internal Roundcube functions can be used inside this cronjob
	 *
	 * @return	void
	 */
	private  function include_rc_iniset()
	{
		if (file_exists($this->doc_root . $this->rc_iniset))
		{
			chdir($this->doc_root);
			require_once $this->doc_root . '/program/include/iniset.php';
		}
		else
		{
			die('Can\'t detect file path correctly! I got this as Roundcubes document root: '. $this->doc_root);
		}
	}

	/**
	 * Get all CardDAV servers
	 *
	 * @return	array	$servers	CardDAV server array with label, url, username, password (encrypted)
	 */
	private function get_carddav_servers()
	{
		$servers	= array();
		$rcmail		= rcmail::get_instance();

		$query = "
			SELECT
				*
			FROM
				".get_table_name('carddav_server')."
		";

		$result = $rcmail->db->query($query);

		while($server = $rcmail->db->fetch_assoc($result))
		{
			$servers[] = $server;
		}

		return $servers;
	}

	/**
	 * Synchronize all available CardDAV servers
	 *
	 * @return	void
	 */
	public function synchronize()
	{
		$servers	= $this->get_carddav_servers();
		$rcmail		= rcmail::get_instance();

		carddav::write_log('CRONJOB: Starting automatic CardDAV synchronization!');

		if (!empty($servers))
		{
			foreach ($servers as $server)
			{
				$rcmail->user->data['user_id'] = $server['user_id'];
				$carddav_addressbook = new carddav_addressbook($server['carddav_server_id'], $server['label'], false);
				$carddav_addressbook->carddav_addressbook_sync($server);
			}
		}

		carddav::write_log('CRONJOB: Automatic CardDAV synchronization finished!');
	}
}

$cronjob = new carddav_synchronization_cronjob();
$cronjob->synchronize();

carddav.php
<?php

/**
 * Include required CardDAV classes
 */
require_once dirname(__FILE__) . '/carddav_backend.php';
require_once dirname(__FILE__) . '/carddav_addressbook.php';

/**
 * Roundcube CardDAV implementation
 *
 * This is a CardDAV implementation for roundcube 0.6 or higher. It allows every user to add
 * multiple CardDAV server in their settings. The CardDAV contacts (vCards) will be synchronized
 * automaticly with their addressbook.
 *
 * @author Christian Putzke <christian.putzke@graviox.de>
 * @copyright Christian Putzke @ Graviox Studios
 * @since 06.09.2011
 * @link http://www.graviox.de/
 * @link https://twitter.com/graviox/
 * @version 0.5.1
 * @license http://www.gnu.org/licenses/agpl.html GNU AGPL v3 or later
 *
 */

class carddav extends rcube_plugin
{
	/**
	 * Roundcube CardDAV version
	 *
	 * @constant string
	 */
	const VERSION = '0.5.3';

	/**
	 * Tasks where the CardDAV plugin is loaded
	 *
	 * @var string
	 */
	public $task = 'settings|addressbook|mail';

	/**
	 * CardDAV addressbook
	 *
	 * @var string
	 */
	protected $carddav_addressbook = 'carddav_addressbook';

	/**
	 * Init CardDAV plugin - register actions, include scripts, load texts, add hooks
	 *
	 * @return	void
	 */
	public function init()
	{
		$rcmail		= rcmail::get_instance();
		$skin_path	= $this->local_skin_path();

		if (!is_dir($skin_path))
		{
			$skin_path = 'skins/default';
		}

		$this->add_texts('localization/', true);
		$this->include_stylesheet($skin_path . '/carddav.css');

		switch ($rcmail->task)
		{
			case 'settings':
				$this->register_action('plugin.carddav-server', array($this, 'carddav_server'));
				$this->register_action('plugin.carddav-server-save', array($this, 'carddav_server_save'));
				$this->register_action('plugin.carddav-server-delete', array($this, 'carddav_server_delete'));
				$this->include_script('carddav_settings.js');
				$this->include_script('jquery.base64.js');
				$this->add_hook('addressbooks_list', array($this, 'get_carddav_addressbook_sources'));
			break;

			case 'addressbook':
				if ($this->carddav_server_available())
				{
					$this->register_action('plugin.carddav-addressbook-sync', array($this, 'carddav_addressbook_sync'));
					$this->include_script('carddav_addressbook.js');
					$this->add_hook('addressbooks_list', array($this, 'get_carddav_addressbook_sources'));
					$this->add_hook('addressbook_get', array($this, 'get_carddav_addressbook'));

					$this->add_button(array(
						'command' => 'plugin.carddav-addressbook-sync',
						'imagepas' => $skin_path . '/sync_pas.png',
						'imageact' => $skin_path . '/sync_act.png',
						'width' => 32,
						'height' => 32,
						'title' => 'carddav.addressbook_sync'
					), 'toolbar');
				}
			break;

			case 'mail':
				if ($this->carddav_server_available())
				{
					$this->add_hook('addressbooks_list', array($this, 'get_carddav_addressbook_sources'));
					$this->add_hook('addressbook_get', array($this, 'get_carddav_addressbook'));

					$sources = (array) $rcmail->config->get('autocomplete_addressbooks', array('sql'));
					$servers = $this->get_carddav_server();

					foreach ($servers as $server)
					{
						if (!in_array($this->carddav_addressbook . $server['carddav_server_id'], $sources))
						{
							$sources[] = $this->carddav_addressbook . $server['carddav_server_id'];
							$rcmail->config->set('autocomplete_addressbooks', $sources);
						}
					}
				}
			break;
		}
	}

	/**
	 * Extend the original local_skin_path method with the default skin path as fallback
	 *
	 * @param	boolean		$include_plugins_directory	Include plugins directory
	 * @return	string		$skin_path					Roundcubes skin path
	 */
	public function local_skin_path($include_plugins_directory = false)
	{
		$skin_path	= parent::local_skin_path();

		if (!is_dir($skin_path))
		{
			$skin_path = 'skins/default';
		}

		if ($include_plugins_directory === true)
		{
			$skin_path = 'plugins/carddav/' . $skin_path;
		}

		return $skin_path;
	}

	/**
	 * Get all CardDAV servers from the current user
	 *
	 * @param	boolean	$carddav_server_id	CardDAV server id to load a single CardDAV server
	 * @return	array						CardDAV server array with label, url, username, password (encrypted)
	 */
	public function get_carddav_server($carddav_server_id = false)
	{
		$servers	= array();
		$rcmail		= rcmail::get_instance();
		$user_id	= $rcmail->user->data['user_id'];

		$query = "
			SELECT
				*
			FROM
				".$rcmail->db->table_name('carddav_server')."
			WHERE
				user_id = ?
			".($carddav_server_id !== false ? " AND carddav_server_id = ?" : null)."
		";

		$result = $rcmail->db->query($query, $user_id, $carddav_server_id);

		while($server = $rcmail->db->fetch_assoc($result))
		{
			$servers[] = $server;
		}

		return $servers;
	}

	/**
	* Render available CardDAV server list in HTML
	*
	* @return	string	HTML rendered CardDAV server list
	*/
	protected function get_carddav_server_list()
	{
		$servers	= $this->get_carddav_server();

		if (!empty($servers))
		{
			$skin_path	= $this->local_skin_path(true);
			$content	= html::div(array('class' => 'carddav_headline'), $this->gettext('settings_server'));
			$table		= new html_table(array(
				'cols'	=> 6,
				'class'	=> 'carddav_server_list'
			));

			$table->add_header(array('width' => '13%'), $this->gettext('settings_label'));
			$table->add_header(array('width' => '36%'), $this->gettext('server'));
			$table->add_header(array('width' => '13%'), $this->gettext('username'));
			$table->add_header(array('width' => '13%'), $this->gettext('password'));
			$table->add_header(array('width' => '13%'), $this->gettext('settings_read_only'));
			$table->add_header(array('width' => '13%'), null);

			foreach ($servers as $server)
			{
				// $rcmail->output->button() seems not to work within ajax requests so we build the button manually
				$delete_submit = '<input
					type="button"
					value="'.$this->gettext('delete').'"
					onclick="return rcmail.command(\'plugin.carddav-server-delete\', \''.$server['carddav_server_id'].'\', this)"
					class="button mainaction"
				/>';

				$table->add(array(), $server['label']);
				$table->add(array(), $server['url']);
				$table->add(array(), $server['username']);
				$table->add(array(), '**********');
				$table->add(array(), ($server['read_only'] ? html::img($skin_path . '/checked.png') : null));
				$table->add(array(), $delete_submit);
			}

			$content .= html::div(array('class' => 'carddav_container'), $table->show());
			return $content;
		}
		else
		{
			return null;
		}
	}

	/**
	 * Get CardDAV addressbook instance
	 *
	 * @param	array	$addressbook	Array with all available addressbooks
	 * @return	array 					Array with all available addressbooks
	 */
	public function get_carddav_addressbook($addressbook)
	{
		$servers = $this->get_carddav_server();

		foreach ($servers as $server)
		{
			if ($addressbook['id'] === $this->carddav_addressbook . $server['carddav_server_id'])
			{
				$addressbook['instance'] = new carddav_addressbook($server['carddav_server_id'], $server['label'], ($server['read_only'] == 1 ? true : false));
			}
		}

		return $addressbook;
	}

	/**
	 * Get CardDAV addressbook source
	 *
	 * @param	array	$addressbook	Array with all available addressbooks sources
	 * @return	array					Array with all available addressbooks sources
	 */
	public function get_carddav_addressbook_sources($addressbook)
	{
		$servers = $this->get_carddav_server();

		foreach ($servers as $server)
		{
			$carddav_addressbook = new carddav_addressbook($server['carddav_server_id'], $server['label'], ($server['read_only'] == 1 ? true : false));

			$addressbook['sources'][$this->carddav_addressbook . $server['carddav_server_id']] = array(
				'id' => $this->carddav_addressbook . $server['carddav_server_id'],
				'name' => $server['label'],
				'readonly' => $carddav_addressbook->readonly,
				'groups' => $carddav_addressbook->groups
			);
		}

		return $addressbook;
	}

	/**
	 * Check if CURL is installed or not
	 *
	 * @return	boolean
	 */
	private function check_curl_installed()
	{
		if (function_exists('curl_init'))
		{
			return true;
		}
		else
		{
			return false;
		}
	}

	/**
	 * Synchronize CardDAV addressbook
	 *
	 * @param	boolean	$carddav_server_id	CardDAV server id to synchronize a single CardDAV server
	 * @param	boolean $ajax				Within a ajax request or not
	 * @return	void
	 */
	public function carddav_addressbook_sync($carddav_server_id = false, $ajax = true)
	{
		$servers	= $this->get_carddav_server();
		$result		= false;

		foreach ($servers as $server)
		{
			if ($carddav_server_id === false || $carddav_server_id == $server['carddav_server_id'])
			{
				$carddav_addressbook = new carddav_addressbook($server['carddav_server_id'], $server['label'], ($server['read_only'] == 1 ? true : false));
				$result = $carddav_addressbook->carddav_addressbook_sync($server);
			}
		}

		if ($ajax === true)
		{
			$rcmail = rcmail::get_instance();

			if ($result === true)
			{
				$rcmail->output->command('plugin.carddav_addressbook_message', array(
					'message' => $this->gettext('addressbook_synced'),
					'check' => true
				));
			}
			else
			{
				$rcmail->output->command('plugin.carddav_addressbook_message', array(
					'message' => $this->gettext('addressbook_sync_failed'),
					'check' => false
				));
			}
		}
	}

	/**
	 * Render CardDAV server settings
	 *
	 * @return	void
	 */
	public function carddav_server()
	{
		$rcmail = rcmail::get_instance();
		$this->register_handler('plugin.body', array($this, 'carddav_server_form'));
		$rcmail->output->set_pagetitle($this->gettext('settings'));
		$rcmail->output->send('plugin');
	}

	/**
	* Check if CardDAV server are available in the local database
	*
	* @return	boolean		If CardDAV-Server are available in the local database return true else false
	*/
	protected function carddav_server_available()
	{
		$rcmail		= rcmail::get_instance();
		$user_id	= $rcmail->user->data['user_id'];

		$query = "
			SELECT
				count(*)
			FROM
				".$rcmail->db->table_name('carddav_server')."
			WHERE
				user_id = ?
		";

		$result = $rcmail->db->query($query, $user_id);

		$count = $rcmail->db->fetch_array($result);
		return ($count[0] > 0);

	}

	/**
	 * Check if it's possible to connect to the CardDAV server
	 *
	 * @return	boolean
	 */
	public function carddav_server_check_connection()
	{
		$url		= rcube_utils::parse_input_value(base64_decode($_POST['_server_url']));
		$username	= rcube_utils::parse_input_value(base64_decode($_POST['_username']));
		$password	= rcube_utils::parse_input_value(base64_decode($_POST['_password']));

		$carddav_backend = new carddav_backend($url);
		$carddav_backend->set_auth($username, $password);

		return $carddav_backend->check_connection();
	}

	/**
	 * Render CardDAV server settings formular and register JavaScript actions
	 *
	 * @return	string	HTML CardDAV server formular
	 */
	public function carddav_server_form()
	{
		$rcmail		= rcmail::get_instance();
		$boxcontent	= null;

		if ($this->check_curl_installed())
		{
			$input_label = new html_inputfield(array(
				'name'		=> '_label',
				'id'		=> '_label',
				'size'		=> '17'
			));

			$input_server_url = new html_inputfield(array(
				'name'		=> '_server_url',
				'id'		=> '_server_url',
				'size'		=> '44'
			));

			$input_username = new html_inputfield(array(
				'name'		=> '_username',
				'id'		=> '_username',
				'size'		=> '17'
			));

			$input_password = new html_passwordfield(array(
				'name'		=> '_password',
				'id'		=> '_password',
				'size'		=> '17'
			));

			$input_read_only = new html_checkbox(array(
				'name'		=> '_read_only',
				'id'		=> '_read_only',
				'value'		=> 1
			));

			$input_submit = $rcmail->output->button(array(
				'command' => 'plugin.carddav-server-save',
				'type' => 'input',
				'class' => 'button mainaction',
				'label' => 'save'
			));

			$table = new html_table(array(
				'cols'	=> 6,
				'class'	=> 'carddav_server_list'
			));

			$table->add_header(array('width' => '13%'), $this->gettext('settings_label'));
			$table->add_header(array('width' => '35%'), $this->gettext('server'));
			$table->add_header(array('width' => '13%'), $this->gettext('username'));
			$table->add_header(array('width' => '13%'), $this->gettext('password'));
			$table->add_header(array('width' => '13%'), $this->gettext('settings_read_only'));
			$table->add_header(array('width' => '13%'), null);

			$table->add(array(), $input_label->show());
			$table->add(array(), $input_server_url->show());
			$table->add(array(), $input_username->show());
			$table->add(array(), $input_password->show());
			$table->add(array(), $input_read_only->show());
			$table->add(array(), $input_submit);

			$boxcontent .= html::div(array('class' => 'carddav_headline'), $this->gettext('settings_server_form'));
			$boxcontent .= html::div(array('class' => 'carddav_container'), $table->show());;
		}
		else
		{
			$rcmail->output->show_message($this->gettext('settings_curl_not_installed'), 'error');
		}

		$output = html::div(
			array('class' => 'box carddav'),
			html::div(array('class' => 'boxtitle'), $this->gettext('settings')).
				html::div(array('class' => 'boxcontent', 'id' => 'carddav_server_list'), $this->get_carddav_server_list()).
				html::div(array('class' => 'boxcontent'), $boxcontent).
				html::div(array('class' => 'boxcontent'), $this->get_carddav_url_list())
		);

		return $output;
	}

	/**
	 * Render a CardDAV server example URL list
	 *
	 * @return	string	$content	HTML CardDAV server example URL list
	 */
	public function get_carddav_url_list()
	{
		$content = null;

		$table = new html_table(array(
			'cols'	=> 2,
			'class'	=> 'carddav_server_list'
		));

		$table->add(array(), 'DAViCal');
		$table->add(array(), 'https://example.com/{resource|principal|username}/{collection}/');

		$table->add(array(), 'Apple Addressbook Server');
		$table->add(array(), 'https://example.com/addressbooks/users/{resource|principal|username}/{collection}/');

		$table->add(array(), 'memotoo');
		$table->add(array(), 'https://sync.memotoo.com/cardDAV/');

		$table->add(array(), 'SabreDAV');
		$table->add(array(), 'https://example.com/addressbooks/{resource|principal|username}/{collection}/');

		$table->add(array(), 'ownCloud');
		$table->add(array(), 'https://example.com/remote.php/carddav/addressbooks/{resource|principal|username}/{collection}/');

		$table->add(array(), 'SOGo');
		$table->add(array(), 'https://example.com/SOGo/dav/{resource|principal|username}/Contacts/{collection}/');

		$content .= html::div(array('class' => 'carddav_headline example_server_list'), $this->gettext('settings_example_server_list'));
		$content .= html::div(array('class' => 'carddav_container'), $table->show());

		return $content;
	}

	/**
	 * Save CardDAV server and execute first CardDAV contact sync
	 *
	 * @return void
	 */
	public function carddav_server_save()
	{
		$rcmail = rcmail::get_instance();

		if ($this->carddav_server_check_connection())
		{
			$user_id	= $rcmail->user->data['user_id'];
			$url		= rcube_utils::parse_input_value(base64_decode($_POST['_server_url']));
			$username	= rcube_utils::parse_input_value(base64_decode($_POST['_username']));
			$password	= rcube_utils::parse_input_value(base64_decode($_POST['_password']));
			$label		= rcube_utils::parse_input_value(base64_decode($_POST['_label']));
			$read_only	= (int) rcube_utils::parse_input_value(base64_decode($_POST['_read_only']));

			$query = "
				INSERT INTO
					".$rcmail->db->table_name('carddav_server')." (user_id, url, username, password, label, read_only)
				VALUES
					(?, ?, ?, ?, ?, ?)
			";

			$rcmail->db->query($query, $user_id, $url, $username, $rcmail->encrypt($password), $label, $read_only);

			if ($rcmail->db->affected_rows())
			{
				$this->carddav_addressbook_sync($rcmail->db->insert_id(), false);

				$rcmail->output->command('plugin.carddav_server_message', array(
					'server_list' => $this->get_carddav_server_list(),
					'message' => $this->gettext('settings_saved'),
					'check' => true
				));
			}
			else
			{
				$rcmail->output->command('plugin.carddav_server_message', array(
					'message' => $this->gettext('settings_save_failed'),
					'check' => false
				));
			}
		}
		else
		{
			$rcmail->output->command('plugin.carddav_server_message', array(
				'message' => $this->gettext('settings_no_connection'),
				'check' => false
			));
		}
	}

	/**
	 * Delete CardDAV server and all related local contacts
	 *
	 * @return	void
	 */
	public function carddav_server_delete()
	{
		$rcmail = rcmail::get_instance();
		$user_id = $rcmail->user->data['user_id'];
		$carddav_server_id = rcube_utils::parse_input_value(base64_decode($_POST['_carddav_server_id']));

		$query = "
			DELETE FROM
				".$rcmail->db->table_name('carddav_server')."
			WHERE
				user_id = ?
			AND
				carddav_server_id = ?
		";

		$rcmail->db->query($query, $user_id, $carddav_server_id);

		if ($rcmail->db->affected_rows())
		{
			$rcmail->output->command('plugin.carddav_server_message', array(
				'server_list' => $this->get_carddav_server_list(),
				'message' => $this->gettext('settings_deleted'),
				'check' => true
			));
		}
		else
		{
			$rcmail->output->command('plugin.carddav_server_message', array(
				'message' => $this->gettext('settings_delete_failed'),
				'check' => false
			));
		}
	}

	/**
	 * Extended write log with pre defined logfile name and add version before the message content
	 *
	 * @param	string	$message	Error log message
	 * @return	void
	 */
	public static function write_log($message)
	{
		rcmail::write_log('CardDAV', 'v' . self::VERSION . ' | ' . $message);
	}
}
